### PR TITLE
Add shell: sh

### PR DESCRIPTION
**Pull request by husjon**
The `sh` reference in [Shell_Bash.ts#L28](https://github.com/Taitava/obsidian-shellcommands/blob/main/src/shells/Shell_Bash.ts#L28) unfortunately does not work, in particular on NixOS which does not use traditional [FHS](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard), as such the only way to spawn a shell without knowing the path is to use `/usr/bin/env bash` or `/usr/bin/env zsh` etc.
Since this was not possible to add quickly, I instead decided to add a separate Shell for `/bin/sh` which is available (also on most if not all Linux distros and MacOS).

In short, this PR moves the `sh` reference to a separate Shell (`Shell_Sh.ts`) to allow it to be selected separately.

### CODE REVIEW

**Comment by Taitava** [comment]
Hi. Thank you for your suggestion!

> The sh reference in [Shell_Bash.ts#L28](https://github.com/Taitava/obsidian-shellcommands/blob/main/src/shells/Shell_Bash.ts#L28) unfortunately does not work

What do you mean by _does not work_? Does it give an error message you could share?

Does the shell work if you select [_Use system default_](https://publish.obsidian.md/shellcommands/Environments/Shells#How+to+know+which+shell+is+used)? The `sh` reference was added in issue #281 and it aimed to make the system default choice work, but I can't really remember how shortly or extensively I tested it. Plus there was some [refactoring done afterwards](https://github.com/Taitava/obsidian-shellcommands/commit/4be365f001325264ffdc7d3e1c0688f293a5d52e) that might have affected it.

---

I'll need to consider this a bit. It might make sense to make `sh` an explicit option like you are suggesting. Or I could create a `Shell_OperatingSystemDefault` class and make it work for any path that comes from the `SHELL` (Linux, macOS) or `ComSpec` (Windows) environment variables. Or do both. I'll think about this.

As it's already possible to define [custom shells](https://publish.obsidian.md/shellcommands/Environments/Custom+shells/Custom+shells) that allow you to [specify the shell's path](https://publish.obsidian.md/shellcommands/Environments/Custom+shells/Settings+for+custom+shells#Executable+binary+file+path) freely, I don't see this as an urgent thing, but I do see that it would be an easier option, especially for new users.

---


For this to be merged, I'd like all the commit messages to be prefixed with an issue number. And as this doesn't have an issue yet, it would be good to go the normal feature request route and create a **discussion** for this, after which I can create an issue if I'll decide to merge this code. I haven't received many pull requests yet, so I haven't really polished up the PR process yet. I guess I'll need to write some instructions to recommend creating a discussion before a pull request 🙂.

**Comment by husjon** [comment]
Hi I might have been a bit too quick when writing the initial PR comment but I'll answer your questions below. :)

---

> What do you mean by _does not work_? Does it give an error message you could share?

The error I get is the following: `Shell command failed to execute. Error: spawn /bin/bash ENOENT` (this is with the original released plugin)

> Does the shell work if you select [_Use system default_](https://publish.obsidian.md/shellcommands/Environments/Shells#How+to+know+which+shell+is+used)?

I've always used the Use System Default option so that unfortunately didn't change anything.

---

> I'll need to consider this a bit. It might make sense to make sh an explicit option like you are suggesting. Or I could create a Shell_OperatingSystemDefault class and make it work for any path that comes from the SHELL (Linux, macOS) or ComSpec (Windows) environment variables. Or do both. I'll think about this.

I think having the possibility for selecting the shell explicitly, especially the usual system defaults (as in shells that usually comes with systems) like sh, bash, zsh and dash is quite a safe choice for at least Linux and Mac (I haven't used Windows in many years so I can't testify for it).
My current shell of choice is Fish, but due to the non-POSIX compliance it's hard to argue for it in a case like this :)

In the case of Bash and the `Shell_Bash.ownedShellBinaries` array, it seem to only be referenced in `BuiltinShell.matchesIdentifier`, this seem to unfortunately only checks if a Shell owns a binary from within the array, i.e `sh`, it would then still be using `/bin/bash` as the shell executable as defined in `Shell_Bash.getBinaryPath()`.

> As it's already possible to define [custom shells](https://publish.obsidian.md/shellcommands/Environments/Custom+shells/Custom+shells) that allow you to [specify the shell's path](https://publish.obsidian.md/shellcommands/Environments/Custom+shells/Settings+for+custom+shells#Executable+binary+file+path) freely, I don't see this as an urgent thing, but I do see that it would be an easier option, especially for new users.

I did see that this was possible but didn't look further into it, I could take a look at how to define one.

One really nice way of handling this would be to allow the shell to be run using `/usr/bin/env <shell of choice>` (e.g: `/usr/bin/env bash`).
This way as long as the shell is part of the environment (specified / overridden by the PATH environment variable), it will choose that executable.

This is for example used in Python Virtual Environments.
The typical shebang of a executable python file usually looks like this:
```python
#!/usr/bin/env python
```

Then when you're using the virtual environment, PATH is prepended with that virtual environments python executable, this has the upside of not touching the System-level python packages.

**Comment by husjon** [comment]
While I was writing the comment above I was also experimenting a bit more using the main branch just to familiarize myself with the flow of the plugin.
In particular how it uses and matches on the `ownedShellBinaries` array.
What I found was that during the shell matching in `getShell` (`ShellFunctions.ts`), the path it identifies with is `/bin/sh` but the executable it's using while spawning it is still `/bin/bash`. (see image below)

The ChildProsess log is in `ShellCommandExecutor.executeShellCommand`, just after the shell has been spawned.

I might just make this PR into a fix for that instead, thoughts?

image

**Comment by husjon** [comment]
So I've played around a bit more, the issue I described in the above comment is still relevant, as in even though `sh` is matched with the Shell_Bash implementation, it is still trying to execute the shell as `/bin/bash` which in my case does not work.
I'm not entirely sure how to most easily split up `getBinaryPath` based on the identified shell, you might have some ideas if it comes to it.

I can however as you mentioned use Custom Shells and point it to `/bin/sh`, this does work as intended.

image


I'm also able to create a custom shell using `/usr/bin/env bash` using the following Shell arguments `-S bash -c {{!shell_command_content}}` (note the `-S` to tell /usr/bin/env to split the following parts as shebang lines)

image



If you do decide to accept the PR as is (to add `sh` as a separate explicit shell), just let me know and I'll correct the commit messages according to how you want them as you mentioned in your original comment. :)

**Comment by Taitava** [comment]
Thank you for your replies! 🙂 I didn't read it all yet as I'm in a bit of hurry now, but I'll come back to this in the coming weeks, if not this week.

> In the case of Bash and the `Shell_Bash.ownedShellBinaries` array, it seem to only be referenced in `BuiltinShell.matchesIdentifier`, this seem to unfortunately only checks if a Shell owns a binary from within the array, i.e `sh`, it would then still be using `/bin/bash` as the shell executable as defined in `Shell_Bash.getBinaryPath()`.

Yes, this is the case. The shell paths are hardcoded, and even the "Use system default" option does not really use the system's default shell path, instead it just tries to match the shell binary's name with the hardcoded paths, and then uses one of them. It's definitely confusing. Right now I can't remember how I came up with this (I could look into this), but it's definitely something that can be improved to really use the system default shell path.

**Comment by Taitava** [comment]
Hi again @husjon , and sorry for the delay! I'll answer the rest of your messages now.

I have thought about this addition of `/bin/sh` as a separate shell selection for Linux and macOS, and yes, let's do it 👍 . And I will probably fix the "Use system default" shell option to actually use the correct path. I haven't decided yet, will these changes happen in the same release or not, but at least the former (using your PR) could be prioritized here and planned for the next version, SC `0.24.0`.

## About `/usr/bin/env <shell of choice>`

> One really nice way of handling this would be to allow the shell to be run using `/usr/bin/env <shell of choice>` (e.g: `/usr/bin/env bash`). This way as long as the shell is part of the environment (specified / overridden by the PATH environment variable), it will choose that executable.
> 
> This is for example used in Python Virtual Environments. The typical shebang of a executable python file usually looks like this:
> 
> ```python
> #!/usr/bin/env python
> ```
> 
> Then when you're using the virtual environment, PATH is prepended with that virtual environments python executable, this has the upside of not touching the System-level python packages.

Hmm, this concept of `/usr/bin/env <shell of choice>` is new to me, but if I understood correctly, this could be a broader way to support different Linux distros that may have Bash, Dash or Zsh installed somewhere else than in `/bin`. Do you know if it works on macOS, too, or only on Linux?

I do like the idea. If I was now deciding what paths to use for the default shells, I would definitely try `/usr/bin/env <bash|dash|zsh>` before (the current) `/bin/<bash|dash|zsh>`.  If `/usr/bin/env` would work for all Linux users (= I don't know this now), I'd like to use it instead of the current one. But I assume your suggestion was to add it as an alternative option? Maybe yes, I'll need to think about how to do that without cluttering up the UI.

(Btw. Do you know if using `/usr/bin/env` affects the performance in any way? I was just thinking that it's one step more before launching the actual shell. But maybe it's unnoticeable.)

Maybe the simplest would be to just duplicate the current Linux shell choices (and macOS, if applicable):

- Rename _Bash_ -> _Bash (/bin/bash)_
- Rename _Dash_ -> _Dash (/bin/dash)_
- Rename _Zsh (Z shell)_ -> _Z shell (/bin/zsh)_
- Add _Bash (/usr/bin/env bash)_
- Add _Dash (/usr/bin/env dash)_
- Add _Z shell (/usr/bin/env zsh)_

This is not a decision yet, just thinking.

## Possibly fixing the _Use system default_ choice

> So I've played around a bit more, the issue I described in the above comment is still relevant, as in even though `sh` is matched with the Shell_Bash implementation, it is still trying to execute the shell as `/bin/bash` which in my case does not work.  
I'm not entirely sure how to most easily split up `getBinaryPath` based on the identified shell, you might have some ideas if it comes to it.

I'll probably look into this myself at some point. Maybe I'll create a class named like `Shell_PlatformDefault` whose `getBinaryPath()` will return whatever happens to be in the `SHELL` or `ComSpec` environment variable.

(I guess one idea I might originally have had with recognizing the shell (the current strange way) is to determine [how special characters in `{{variable}}` values should be escaped](https://publish.obsidian.md/shellcommands/Variables/Escaping+special+characters+in+variable+values), i.e. what character to use for escaping (currently backslash `\` or backtick `` ` ``). Now I think that maybe I don't have to know each and all shells out there, I can just assume that on Linux and macOS, the escaping can use backlash `\` , and on Windows PowerShell, it can use backtick `` ` ``. If there's any shell that uses different escaping, I can always add special cases later if someone reports problems. To determine the escaping system, it was needed to pick a `Shell_*` class that generally suits the user's shell, but ending up also using the hardcoded binary path seems to be an accident, and the real default path should have been used right from the beginning.)

- [ ] I need to create an issue about this for myself. Like I mentioned in the beginning of this comment, I haven't decided a milestone (SC version) for this yet.

## Finishing up the addition of `/bin/sh` shell

- I have created issue #440 for this. (I didn't create a discussion like features would normally have, as this PR can exceptionally serve as a discussion. 🙂)
- Now you can edit the commit messages of this PR and add the `#440: ` prefix to include the issue reference in them.
- If you want, you can also edit AUTHORS.md in this pull request and add your details as a contributor. This is completely optional.

Thank you very much for your time and effort on describing this problem and coming up with a solution! 🙏😎

**Comment by husjon** [comment]
Hi and no worries about the delay, I was able to get unblocked by creating a custom shell. :)

> Hmm, this concept of /usr/bin/env <shell of choice> is new to me, but if I understood correctly, this could be a broader way to support different Linux distros that may have Bash, Dash or Zsh installed somewhere else than in /bin. Do you know if it works on macOS, too, or only on Linux?

I did a quick search yesterday and found a few posts saying that it is available on MacOS too (one from 2017 https://apple.stackexchange.com/questions/297533/does-macos-have-usr-bin-env which also links the manpage for it), it is a different version (BSD vs GNU) but the end result is the same, which in this context is to look up the shell to run the commands as.

I can't speak for all non-FHS systems, but in in my case with NixOS it tries to bridge the gap with at least providing some bare minimum known paths, like `/bin/sh` and `/usr/bin/env`.
Everything else is handled with environment variables.
The 2 paths mentioned above would also be needed to be exposed on other non-FHS systems if they should have any chance of running "normal" apps that are not build explicitly for said system. :)

> (Btw. Do you know if using /usr/bin/env affects the performance in any way? I was just thinking that it's one step more before launching the actual shell. But maybe it's unnoticeable.)

As for performance, it would call it negligible.
### Single execution
* `/bin/bash`: 0.003 seconds
* `/usr/bin/env`: 0.004 seconds
image

### 1000 runs
* `/bin/bash`: 2.914 seconds
* `/usr/bin/env`: 3.483 seconds
image


> Maybe the simplest would be to just duplicate the current Linux shell choices (and macOS, if applicable)

To reduce the need for duplication, it would be nice if for example the Bash shell (and the others) tried `/usr/bin/env bash` first, then `/bin/bash` as a fallback.
This way it would let the executables in the users environment take precedence over system executables and only use those as fallbacks.
I am however not sure how this would work on Windows since I haven't used it in quite a few years (it should work fine on MacOS).
If it is too hard to implement, adding duplicate entries as you showed also works. :)

> I guess one idea I might originally have had with recognizing the shell (the current strange way) is to determine [how special characters in {{variable}} values should be escaped](https://publish.obsidian.md/shellcommands/Variables/Escaping+special+characters+in+variable+values)

I found this out the other day as I was trying to use the provided variables (`{{vault_path}}`) and I found the values to be pre-escaped, I'm used to always wrapping variables which expands (especially ones that expands to paths) in quotation marks (`"`) so that spaces etc are kepts as-is while keeping it human-readable.
As in, instead of `/home/user/my\ special\ file\ with\ spaces.md`,  
 it would be `"/home/user/my special file with spaces.md"`.

I think this would also work for Windows systems as in if you wrap the variable in normal double quotation marks it should keep the string as is, but I might be wrong. :)

On Linux `"$MY_VAR"` would expand the MY_VAR variable, while `'$MY_VAR'` would not.
This would also be the case on MacOS.


> Finishing up the addition of /bin/sh shell

I'll go through and rebase and update the commit messages with the issue number shortly after finishing this message.

Happy to help and thanks you too. :)

**Comment by Taitava** [comment]
Hi @husjon and sorry for the delay again!

> > Maybe the simplest would be to just duplicate the current Linux shell choices (and macOS, if applicable)
> 
> To reduce the need for duplication, it would be nice if for example the Bash shell (and the others) tried `/usr/bin/env bash` first, then `/bin/bash` as a fallback. This way it would let the executables in the users environment take precedence over system executables and only use those as fallbacks.

Your idea seems good in a way that it would keep the configuration simpler for user. However, I'd need to solve the following problems:
- Currently, Bash, Dash and Zsh shell choices are stored in data.json as absolute paths, i.e. `/bin/bash`, `/bin/dash` and `/bin/zsh` (they are kind of "constant values", i.e. non-customisable in the config file). If I'd change their behavior so that `/bin/bash` (defined in data.json) would in some cases use `/usr/bin/env bash` in practice, it might be misleading. Maybe it could be solved by migrating the configuration values to something more ambiguous, e.g. just `bash`, but maybe I wouldn't opt for that.
- I'd like to improve the current user interface so that the settings panel would show the absolute path to the shell binary when picking a shell in a dropdown. If there was some automation used to determine the shell's path, the UI would not be able to show a path until at least one shell command has been executed. Or maybe it could show both paths, e.g. _Bash (/usr/bin/env bash or /bin/bash)_.
- For users who currently have both `/usr/bin/env bash` and `/bin/bash` present, modifying the current `/bin/bash` choice would suddenly switch the shell binary. Could the order of the tried paths be reversed? First try `/bin/bash` and if not found, then try `/usr/bin/env bash`? (I used `bash` here as an example, but the same applies to `dash` and `zsh`, too).

To sum it up, I need to think about this idea more in order to decide whether I'll implement this kind of automation, or the separate choices.

(Also one thing that came to my mind is that should `/usr/bin/bash` be one choice the path trier could try? I do not know the difference between `/bin` and `/usr/bin`, so I don't know can there be a situation where `/usr/bin/bash` exists but `/bin/bash` doesn't?)

> I am however not sure how this would work on Windows since I haven't used it in quite a few years (it should work fine on MacOS).

If this path trying feature would be implemented, I'd do it so that's its only tried on Linux and macOS.

-----------

> I found this out the other day as I was trying to use the provided variables (`{{vault_path}}`) and I found the values to be pre-escaped, I'm used to always wrapping variables which expands (especially ones that expands to paths) in quotation marks (`"`) so that spaces etc are kepts as-is while keeping it human-readable.
> As in, instead of `/home/user/my\ special\ file\ with\ spaces.md`,
> it would be `"/home/user/my special file with spaces.md"`.

Yeah, wrapping in double quotes makes sense for variables that always return file/folder paths, and the result is more readable. Maybe you already know, but you can prevent the escaping by using an exclamation mark, e.g. `{{!vault_path}}` when wrapping the path in double quotes. That works ok for path variables, because their values will never contain double quotes, so they cannot accidentally break out from the wrapping double quotes. However, I wouldn't use this for other variables (e.g. `{{selection}}`) where the content is more unpredictable.

---------------

> I'll go through and rebase and update the commit messages with the issue number shortly after finishing this message. I also added myself to AUTHORS.md.
> 
> Happy to help and thanks to you too. :)

Thank you a lot for your contribution! 😎 I'll do some small testing on my Linux machine some day. By looking at the code, I didn't spot any problems so far. 👍 

If the test is ok, I'll create a new branch `0.24.0` and merge this there. I can't promise when I will release `0.24.0` as it takes quite a lot of time for me to run through [my test vault](https://github.com/Taitava/obsidian-shellcommands-test-suite) that I do before all releases. (I definitely need to automate it more at some point 😅 ).

**Comment by husjon** [comment]
Hi @taitava and no worries at all.

First of, this is going to be a wall of text, I've tried to split it up in sections by point. 😅

### Currently, Bash, Dash and Zsh shell choices are stored in data.json as absolute paths...
I had to take a look at how the `data.json` file was structured to see what you meant. :)
What I can see are the following variations (everything omitted are the same across each variation).

1. Use default (Bash)
Here, shells are not populated
```json
{
  "shell_commands": [
    {
      "id": "otkctosvle",
      "platform_specific_commands": { "default": "echo 'Hello, World!'" },
      "shells": {},
    }
  ]
}
```

2. Bash (manually selected)
The specific shell path is hardcoded
```json
{
  "shell_commands": [
    {
      "id": "otkctosvle",
      "platform_specific_commands": { "default": "echo 'Hello, World!'" },
      "shells": {
        "linux": "/bin/bash"
      }
    }
  ]
}
```

3. Custom shell
When selecting a custom shell, the ID for that shell is referenced.
```json
{
  "shell_commands": [
    {
      "id": "otkctosvle",
      "platform_specific_commands": { "default": "echo 'Hello, World!'" },
      "shells": {
        "linux": "5ua0sgze58"
      }
    }
  ]
}
```

If all 3 referenced a shell by an ID, I think the automatic selection could be handled from within each shell itself or the `BaseShell` class.
For builtin shells, I think the IDs could be `bash`, `dash`, `zsh` to make it easily readable in `data.json`, then keep the IDs as they are today for custom shells.

This would however need a pre-initialization step so that each shell could look up if that particular shell path is available.

Adding a method to `BaseShell` named `lookupShellPath` or something along those lines, we could utilize `whereis` (Linux / Mac) or `where` (Windows) to look up the paths for the shells.
Using the following for inspiration: https://stackoverflow.com/a/34953561

```ts
import { spawnSync } from "child_process";
import { platform } from "os";

export abstract class BuiltinShell extends Shell {
    public lookupShellPath(shell:string): string[] {
        var isWin = platform().indexOf("win") > -1;
        var where = isWin ? "where" : "whereis";

        const command = spawnSync(where, [shell]);  // replace `shell` with shell / executable of choice
        const stdoutString = String.fromCharCode.apply(null, command.stdout);  // `stdout` is a `Uint8Array` so we need to convert it to a string
        const shellPaths = stdoutString
            .split(isWin ? '\n' : ' ') // Windows is newline separated, while Linux is space separated (Mac should also be space separated)
            .slice(isWin ? 0 : 1) // Linux prefixes the output with the first argument hence we need to skip past the first element
            .map((path: string) => path.trim()); // trim each path as they can contain dangling newlines

        return shellPaths;  // Example using Arch Linux [ "/usr/bin/bash", "/usr/lib/bash", "/usr/include/bash" ]
    }
}
```

The output of these commands are the following on different OSes:
```
# Ubuntu (22.04 LTS)
$ whereis bash
bash: /usr/bin/bash

# Arch Linux
$ whereis bash
bash: /usr/bin/bash /usr/lib/bash /usr/include/bash

# NixOS
$ whereis bash
bash: /nix/store/nsqjbhwj7v083qlx8xlrkmcgbni724hc-user-environment/bin/bash /nix/store/jlr7bdkyl10caxc8ps0x8z7pphca807x-system-path/bin/bash

# Windows 11
C:\>where *.exe
C:\Windows\System32\alg.exe
C:\Windows\System32\at.exe
... redacted for brevity ...

# non-existing command on linux / mac
$ whereis non-existing
non-existing:

# non-existing command on Windows
$ where non-existing-command.exe
INFO: Could not find files for the given pattern(s).
```

### I'd like to improve the current user interface so that the settings panel would show the absolute path...
This might be easier with the example above since the paths would be known on startup / initialization.
I'm not sure how I would present them, but as long as the paths are known that should make it a bit easier to figure out over time.


### For users who currently have both /usr/bin/env bash and /bin/bash present...
This would be the same as when someone opens a shell / terminal and types for example `bash` in the terminal.
One of the first thing that happens is that the executable is looked up using the users `PATH` variable and then executes the first one it finds.
The only way to run for example `/bin/bash` specifically is for the user to either explicitly type `/bin/bash` or by having `/bin` as the first path in the `PATH` variable. :)

One quick way to test this is if you create a folder named `bin` in your home directory, then prefix `PATH` with this new path.
```sh
export PATH=~/bin:$PATH
```
If you now were to create a script which had the name `bash` under `~/bin`, running the command `bash` would now execute this new script. :)

Alternatively, if you run `unset PATH`, the only way to run bash from the current shell is to run it using the full path.



### Also one thing that came to my mind is that should /usr/bin/bash be one choice the path trier could try...
Using the proposal above, it will:
1. Iterate through all directories that is listed in the `PATH` variable 
2. Add the full path of the executable named `bash` from each directory.

So, if the user has `/bin`, `/usr/bin` and `/usr/local/bin` in their PATH in that order and `bash` **only** exist in the latter, it would only return: `/usr/local/bin/bash`


### Final thoughts
With this, it opens up the possibility to hide shells that are not available on different systems.
As an example `Dash` is not on my system and as such should be disabled in the dropdown (maybe also with a note about it not being available).
image

Regarding presenting the available shells and their paths could be like this, using option groups `optgroup`
Here's a quick mockup with shells having more than one path.
image
```html
<select class="dropdown">
  <option value="default">Use system default (bash)</option>

  <optgroup label="Bash">
    <option value="/bin/bash">/bin/bash</option>
    <option value="/bin/bash">/usr/bin/bash</option>
  </optgroup>
  <optgroup label="Dash">
    <option value="/bin/dash">/bin/dash</option>
  </optgroup>
  <optgroup label="Zsh (Z shell)">
    <option value="/bin/zsh">/bin/zsh</option>
    <option value="/bin/bash">/usr/bin/zsh</option>
    <option value="/bin/bash">/usr/local/bin/zsh</option>
  </optgroup>
  <optgroup label="Sh">
    <option value="/bin/sh">/bin/sh</option>
  </optgroup>

  <optgroup label="Custom Shells">
    <option value="5ua0sgze58">/bin/sh</option>
  </optgroup>
</select>
```
PS: Custom Shells should show the Shell name rather than the path (or maybe both) 

---


Regarding the double-quotes did not know about it but I'll be using it going forward.


---

Once again, no worries at all about the delay, this is a massive change in an already big project, also the current implementation I'm using (custom shells) works absolutely fine.


If you should need any feedback / help or if there might be something I missed, just let me know. :)
